### PR TITLE
[Feature] Support PPYOLOE inference

### PR DESCRIPTION
**Pull request by Nioolek**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

add ppyoloe and ppyoloeplus

## Modification

Please briefly describe what modification is made in this PR.

The code can reproduce the coco mAP result for ppyoloe_plus s model.
image

Official ppyoloe_plus mAP result:
image

The gap of 0.1 is derived from the inference difference between the paddle and pytorch frameworks, and official ppyoloe plus filter imgs with no gt during test, but we dont.

### CODE REVIEW

**Comment by hhaAndroid** [comment]
@Nioolek Please fix the conflicting files

**Review by PeterH0323 — CHANGES_REQUESTED** [review]

File: configs/ppyoloe/ppyoloe_plus_m_fast_8xb8-80e_coco.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+model = dict(
+    backbone=dict(
+        type='CSPResNet',
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor),
+    neck=dict(
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor,
+    ),
+    bbox_head=dict(head_module=dict(widen_factor=widen_factor), ))
```
Comment (PeterH0323) [inline configs/ppyoloe/ppyoloe_plus_m_fast_8xb8-80e_coco.py:16]:
```suggestion
    bbox_head=dict(head_module=dict(widen_factor=widen_factor)))
```

File: configs/ppyoloe/ppyoloe_plus_x_fast_8xb8-80e_coco.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+model = dict(
+    backbone=dict(
+        type='CSPResNet',
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor),
+    neck=dict(
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor,
+    ),
+    bbox_head=dict(head_module=dict(widen_factor=widen_factor), ))
```
Comment (PeterH0323) [inline configs/ppyoloe/ppyoloe_plus_x_fast_8xb8-80e_coco.py:16]:
```suggestion
    bbox_head=dict(head_module=dict(widen_factor=widen_factor)))
```

File: mmyolo/models/backbones/csp_resnet.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            BasicBlock,
+            in_channels,
+            out_channels,
+            num_blocks,
+            2,
+            norm_cfg=self.norm_cfg,
+            act_cfg=self.act_cfg,
+            use_alpha=self.use_alpha)
+        stage.append(cspres_layer)
+        return stage
```
Comment (PeterH0323) [inline mmyolo/models/backbones/csp_resnet.py:244]: 这里可否直接写成
```python
return [cspres_layer]
```

File: mmyolo/models/dense_heads/ppyoloe_head.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            1,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+
+        self._init_weights()
+
+    def _init_weights(self):
+        nn.init.normal_(self.fc.weight, mean=0, std=0.001)
+
+    def forward(self, feat, avg_feat):
```
Comment (PeterH0323) [inline mmyolo/models/dense_heads/ppyoloe_head.py:48]: type hint

File: mmyolo/models/backbones/csp_resnet.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        self.conv2 = RepVGGBlock(
+            output_channels,
+            output_channels,
+            alpha=use_alpha,
+            act_cfg=act_cfg,
+            norm_cfg=norm_cfg,
+            mode='ppyoloe')
+        self.shortcut = shortcut
+
+    def forward(self, x):
```
Comment (PeterH0323) [inline mmyolo/models/backbones/csp_resnet.py:120]: type hint

File: mmyolo/models/backbones/csp_resnet.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+            self.attn = None
+
+        self.conv3 = ConvModule(
+            middle_channels,
+            output_channels,
+            1,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+
+    def forward(self, x):
```
Comment (PeterH0323) [inline mmyolo/models/backbones/csp_resnet.py:76]: type hint

File: mmyolo/models/layers/yolo_bricks.py [hunk 7]
```diff
@@ -329,3 +353,25 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+
+    def __init__(self, channels, act_cfg=dict(type='HSigmoid')):
+        super().__init__()
+        assert isinstance(act_cfg, dict)
+        self.fc = ConvModule(channels, channels, 1, act_cfg=None)
+
+        act_cfg_ = act_cfg.copy()  # type: ignore
+        self.activate = build_activation_layer(act_cfg_)
+
+    def forward(self, x):
```
Comment (PeterH0323) [inline mmyolo/models/layers/yolo_bricks.py:374]: type hint

File: mmyolo/models/necks/ppyoloe_customcsppan.py [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+            self.pool.append(pool)
+        self.conv = ConvModule(
+            input_channels,
+            output_channels,
+            kernel_size,
+            padding=kernel_size // 2,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+
+    def forward(self, x):
```
Comment (PeterH0323) [inline mmyolo/models/necks/ppyoloe_customcsppan.py:195]: type hint

**Review by PeterH0323 — CHANGES_REQUESTED** [review]

File: mmyolo/datasets/ppyoloe_coco.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import List
+
+from mmdet.datasets import CocoDataset
+
+from ..registry import DATASETS
+
+
+@DATASETS.register_module()
+class PPYOLOECocoDataset(CocoDataset):
```
Comment (hhaAndroid) [inline mmyolo/datasets/ppyoloe_coco.py:10]: unnecessary

File: configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    backbone=dict(
+        type='CSPResNet',
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor,
+        norm_cfg=dict(type='BN', momentum=0.1, eps=1e-5),
+        act_cfg=dict(type='Swish'),
+        use_large_stem=True,
+        use_alpha=True),
+    neck=dict(
+        type='PPYOLOECustomCSPPAN',
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py:39]: PPYOLOECustomCSPPAN ->PPYOLOECSPPAN

File: configs/ppyoloe/ppyoloe_plus_l_fast_8xb8-80e_coco.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
+_base_ = './ppyoloe_plus_s_syncbn_fast_8xb32-300e_coco.py'
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_l_fast_8xb8-80e_coco.py:1]:
```suggestion
_base_ = './ppyoloe_plus_s_fast_8xb32-300e_coco.py'
```

Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_l_fast_8xb8-80e_coco.py:1]: Let's put all the configurations of the non-plus version, but we will not to train.

File: configs/ppyoloe/ppyoloe_plus_l_fast_8xb8-80e_coco.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+model = dict(
+    backbone=dict(
+        type='CSPResNet',
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor),
+    neck=dict(
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor,
+    ),
+    bbox_head=dict(head_module=dict(widen_factor=widen_factor), ))
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_l_fast_8xb8-80e_coco.py:16]:
```suggestion
    bbox_head=dict(head_module=dict(widen_factor=widen_factor)))
```

File: configs/ppyoloe/ppyoloe_plus_l_fast_8xb8-80e_coco.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
+_base_ = './ppyoloe_plus_s_syncbn_fast_8xb32-300e_coco.py'
+
+deepen_factor = 1.0
+widen_factor = 1.0
+train_batch_size_per_gpu = 8
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_l_fast_8xb8-80e_coco.py:5]: train_batch_size_per_gpu invalid

File: configs/ppyoloe/ppyoloe_plus_m_fast_8xb8-80e_coco.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
+_base_ = './ppyoloe_plus_s_syncbn_fast_8xb32-300e_coco.py'
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_m_fast_8xb8-80e_coco.py:1]:
```suggestion
_base_ = './ppyoloe_plus_s_fast_8xb32-300e_coco.py'
```

File: configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
+_base_ = '../_base_/default_runtime.py'
+
+# dataset settings
+data_root = 'data/coco/'
+dataset_type = 'mmdet.CocoDataset'
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py:5]: YOLOv5COCOxxx

File: configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+val_num_workers = 2
+
+# persistent_workers must be False if num_workers is 0.
+persistent_workers = True
+
+strides = [8, 16, 32]
+
+# single-scale training is recommended to
+# be turned on, which can speed up training.
+env_cfg = dict(cudnn_benchmark=True)
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py:25]: delete

File: configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+model = dict(
+    type='YOLODetector',
+    data_preprocessor=dict(
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py:29]: The fast version is not written in this way

File: configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        type='mmdet.DetDataPreprocessor',
+        mean=[0., 0., 0.],
+        std=[255., 255., 255.],
+        bgr_to_rgb=True),
+    backbone=dict(
+        type='CSPResNet',
+        deepen_factor=deepen_factor,
+        widen_factor=widen_factor,
+        norm_cfg=dict(type='BN', momentum=0.1, eps=1e-5),
+        act_cfg=dict(type='Swish'),
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py:39]:
```suggestion
        act_cfg=dict(type='nn.SiLu'),
```

File: mmyolo/models/backbones/csp_resnet.py [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+from mmdet.utils import ConfigType, OptMultiConfig
+
+from mmyolo.models.backbones import BaseBackbone
+from mmyolo.models.layers.yolo_bricks import EffectiveSELayer, RepVGGBlock
+from mmyolo.models.utils import make_divisible, make_round
+from mmyolo.registry import MODELS
+
+
+class CSPResStage(nn.Module):
+    """PPYOLOE Backbone Stage."""
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:16]: docstr too little information

File: mmyolo/models/backbones/csp_resnet.py [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+                input_channels,
+                middle_channels,
+                3,
+                stride=2,
+                padding=1,
+                norm_cfg=norm_cfg,
+                act_cfg=act_cfg)
+        else:
+            self.conv_down = None
+        self.conv1 = ConvModule(
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:42]: If there is a specific meaning, it is best not to use 123 instead

File: mmyolo/models/backbones/csp_resnet.py [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+        self.blocks = nn.Sequential(*[
+            block_fn(
+                middle_channels // 2,
+                middle_channels // 2,
+                norm_cfg=norm_cfg,
+                act_cfg=act_cfg,
+                shortcut=True,
+                use_alpha=use_alpha) for i in range(num_layer)
+        ])
+        if use_attn:
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:63]: use_effective_se

File: mmyolo/models/backbones/csp_resnet.py [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+            Defaults to False.
+        use_alpha (bool): Whether to use alpha in `RepVGGBlock`. PPYOLOE do
+            not use, but PPYOLOE+ use.
+            Defaults to False.
+    """
+    # From left to right:
+    # in_channels, out_channels, num_blocks, add_identity, use_spp
+    arch_settings = {
+        'P5': [[64, 128, 3, True, False], [128, 256, 6, True, False],
+               [256, 512, 6, True, False], [512, 1024, 3, True, False]]
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:169]: Can the user set use_spp=True? Don't expose parameters if not allowed

File: mmyolo/models/backbones/csp_resnet.py [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+            norm_eval=norm_eval,
+            init_cfg=init_cfg)
+
+    def build_stem_layer(self) -> nn.Module:
+        """Build a stem layer."""
+        if self.use_large_stem:
+            stem = nn.Sequential(
+                ConvModule(
+                    self.input_channels,
+                    make_divisible(self.arch_setting[0][0], self.widen_factor)
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:212]: int() replace

File: mmyolo/models/necks/yolov5_pafpn.py [hunk 24]
```diff
@@ -35,7 +35,7 @@ class YOLOv5PAFPN(BaseYOLONeck):
 
     def __init__(self,
                  in_channels: List[int],
-                 out_channels: int,
+                 out_channels: List[int],
```
Comment (hhaAndroid) [inline mmyolo/models/necks/yolov5_pafpn.py:38]:
```suggestion
                 out_channels: Union[List[int],int],
```

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 25]
```diff
@@ -0,0 +1,373 @@
...
+from mmcv.cnn import ConvModule
+from mmdet.utils import ConfigType, OptMultiConfig
+
+from mmyolo.models.layers.yolo_bricks import RepVGGBlock
+from mmyolo.models.necks import BaseYOLONeck
+from mmyolo.models.utils import make_divisible, make_round
+from mmyolo.registry import MODELS
+
+
+class SPP(nn.Module):
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:15]: why not put it to the backbone?

Comment (Nioolek) [reply mmyolo/models/necks/ppyoloe_csppan.py:15]: PPYOLOE only use `spp` in neck

File: configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+    dict(
+        type='LoadImageFromFile',
+        file_client_args={{_base_.file_client_args}}),
+    dict(
+        type='mmdet.FixShapeResize',
+        width=img_scale[1],
+        height=img_scale[0],
+        keep_ratio=False,
+        interpolation='bicubic'),
+    dict(type='LoadAnnotations', with_bbox=True),
```
Comment (hhaAndroid) [inline configs/ppyoloe/ppyoloe_plus_s_fast_8xb8-80e_coco.py:80]: dict(type='LoadAnnotations', with_bbox=True, _scope_='mmdet'),

File: mmyolo/models/backbones/csp_resnet.py [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+            middle_channels,
+            middle_channels // 2,
+            1,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+
+        block_cfg = block_cfg.copy()
+        block_cfg['in_channels'] = middle_channels // 2
+        block_cfg['out_channels'] = middle_channels // 2
+        block_cfg['norm_cfg'] = norm_cfg
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:134]: block_cfg.setdefault('out_channels, norm_cfg)
block_cfg.setdefault('act_cfg, act_cfg)

File: mmyolo/models/backbones/csp_resnet.py [hunk 28]
```diff
@@ -0,0 +1,316 @@
...
+        block_cfg['act_cfg'] = act_cfg
+
+        self.blocks = nn.Sequential(
+            *[MODELS.build(block_cfg) for i in range(num_block)])
+        if use_effective_se:
+            self.attn = EffectiveSELayer(
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:140]: 也改成可注册模式？

Comment (hhaAndroid) [reply mmyolo/models/backbones/csp_resnet.py:140]: 这样就不需要 use_efff_se 这个参数了

File: mmyolo/models/backbones/csp_resnet.py [hunk 29]
```diff
@@ -0,0 +1,316 @@
...
+        """
+        y = self.conv1(x)
+        y = self.conv2(y)
+        if self.shortcut:
+            return x + y
+        else:
+            return y
+
+
+class CSPResStage(nn.Module):
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:75]: 如果可以和 neck 的 CSPStage 合并，我建议放到 layer 文件里面去

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 30]
```diff
@@ -0,0 +1,320 @@
...
+        self.convs = self.build_convs_layer(block_cfg, middle_channels,
+                                            num_block)
+        self.conv3 = ConvModule(
+            middle_channels * 2,
+            out_channels,
+            1,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+
+    def build_convs_layer(self, block_cfg: ConfigType, middle_channels: int,
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:71]:
```suggestion
    def build_block_layer(self, block_cfg: ConfigType, middle_channels: int,
```

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 31]
```diff
@@ -0,0 +1,320 @@
...
+            1,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+        self.conv2 = ConvModule(
+            in_channels,
+            middle_channels,
+            1,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+        self.convs = self.build_convs_layer(block_cfg, middle_channels,
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:62]: self.blocks

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 32]
```diff
@@ -0,0 +1,320 @@
...
+        use_drop_block (bool): Whether to use `DropBlock` after `CSPStage`.
+            Defaults to False.
+        drop_block_cfg (dict): If use_drop_block is True, build drop_block
+            by this cfg. Defaults to dict(type='mmdet.DropBlock',drop_prob=0.1,
+            block_size=3, warm_iters=0)
+    """
+
+    def __init__(self,
+                 in_channels: List[int] = [256, 512, 1024],
+                 out_channels: Union[int, List[int]] = [256, 512, 1024],
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:150]: 如果你类不支持单独一个 Int，那就不要写 union

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 33]
```diff
@@ -0,0 +1,320 @@
...
+                     type='PPYOLOEBasicBlock', shortcut=False,
+                     use_alpha=False),
+                 norm_cfg: ConfigType = dict(
+                     type='BN', momentum=0.1, eps=1e-5),
+                 act_cfg: ConfigType = dict(type='SiLU', inplace=True),
+                 init_cfg: OptMultiConfig = None,
+                 num_stage: int = 1,
+                 num_block: int = 3,
+                 use_spp: bool = False,
+                 use_drop_block: bool = False,
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:164]: 不需要两个参数，drop_block_cfg 支持 none 参数jike

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 34]
```diff
@@ -0,0 +1,320 @@
...
+
+        Args:
+            idx (int): layer idx.
+
+        Returns:
+            nn.Module: The reduce layer.
+        """
+        if idx == len(self.in_channels) - 1:
+            # fpn_stage
+            channels = [self.in_channels[idx]
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:202]: 这个很奇怪

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 35]
```diff
@@ -0,0 +1,320 @@
...
+                 widen_factor: float = 1.0,
+                 freeze_all: bool = False,
+                 block_cfg: ConfigType = dict(
+                     type='PPYOLOEBasicBlock', shortcut=False,
+                     use_alpha=False),
+                 norm_cfg: ConfigType = dict(
+                     type='BN', momentum=0.1, eps=1e-5),
+                 act_cfg: ConfigType = dict(type='SiLU', inplace=True),
+                 init_cfg: OptMultiConfig = None,
+                 num_stage: int = 1,
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:161]: num_csplayer

File: mmyolo/models/backbones/csp_resnet.py (hunk 29 above)
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:75]: CSPResStage -> CSPResLayer

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 36]
```diff
@@ -0,0 +1,320 @@
...
+        y1 = self.conv1(x)
+        y2 = self.conv2(x)
+        y2 = self.convs(y2)
+        y = torch.cat([y1, y2], axis=1)
+        y = self.conv3(y)
+        return y
+
+
+@MODELS.register_module()
+class PPYOLOECSPPAN(BaseYOLONeck):
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:115]:
```suggestion
class PPYOLOECSPPAFPN(BaseYOLONeck):
```

File: mmyolo/models/necks/ppyoloe_csppan.py [hunk 37]
```diff
@@ -0,0 +1,320 @@
...
+            block_size=3, warm_iters=0)
+    """
+
+    def __init__(self,
+                 in_channels: List[int] = [256, 512, 1024],
+                 out_channels: Union[int, List[int]] = [256, 512, 1024],
+                 deepen_factor: float = 1.0,
+                 widen_factor: float = 1.0,
+                 freeze_all: bool = False,
+                 block_cfg: ConfigType = dict(
```
Comment (hhaAndroid) [inline mmyolo/models/necks/ppyoloe_csppan.py:154]: num_block 参数放到 blck_cfg前面，放一起，比较好理解

File: mmyolo/models/backbones/csp_resnet.py [hunk 38]
```diff
@@ -0,0 +1,169 @@
...
+                 out_indices: Tuple[int] = (2, 3, 4),
+                 frozen_stages: int = -1,
+                 plugins: Union[dict, List[dict]] = None,
+                 arch_ovewrite: dict = None,
+                 block_cfg: ConfigType = dict(
+                     type='PPYOLOEBasicBlock', shortcut=True, use_alpha=True),
+                 norm_cfg: ConfigType = dict(
+                     type='BN', momentum=0.1, eps=1e-5),
+                 act_cfg: ConfigType = dict(type='SiLU', inplace=True),
+                 effective_se_cfg: ConfigType = dict(
```
Comment (hhaAndroid) [inline mmyolo/models/backbones/csp_resnet.py:72]: attention_cfg

File: mmyolo/models/dense_heads/ppyoloe_head.py [hunk 39]
```diff
@@ -0,0 +1,235 @@
...
+            self.in_channels = [int(i * widen_factor) for i in in_channels]
+
+        self._init_layers()
+
+    def init_weights(self, prior_prob=0.01):
+        """Initialize the weight and bias of PPYOLOE head."""
+        for conv in self.cls_preds:
+            b = conv.bias.view(-1, )
+            b.data.fill_(-math.log((1 - prior_prob) / prior_prob))
+            conv.bias = nn.Parameter(b.view(-1), requires_grad=True)
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/ppyoloe_head.py:77]: openmmlab 不支持这种写法好像，你可以看下yolov5 写法

File: mmyolo/models/dense_heads/ppyoloe_head.py [hunk 40]
```diff
@@ -0,0 +1,235 @@
...
+            conv.weight = nn.Parameter(w, requires_grad=True)
+
+    def _init_layers(self):
+        """initialize conv layers in PPYOLOE head."""
+        self.cls_preds = nn.ModuleList()
+        self.reg_preds = nn.ModuleList()
+        self.cls_stems = nn.ModuleList()
+        self.reg_stems = nn.ModuleList()
+
+        for in_c in self.in_channels:
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/ppyoloe_head.py:97]: in_c -> in_channel

File: mmyolo/models/dense_heads/ppyoloe_head.py [hunk 41]
```diff
@@ -0,0 +1,235 @@
...
+            self.cls_stems.append(
+                ESEAttn(in_c, norm_cfg=self.norm_cfg, act_cfg=self.act_cfg))
+            self.reg_stems.append(
+                ESEAttn(in_c, norm_cfg=self.norm_cfg, act_cfg=self.act_cfg))
+
+        for in_c in self.in_channels:
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/ppyoloe_head.py:103]: in_c -> in_channel

File: mmyolo/models/layers/yolo_bricks.py [hunk 42]
```diff
@@ -336,6 +365,83 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        """Forward process
+         Args:
+             x (Tensor): The input tensor.
+         """
+        x_se = x.mean((2, 3), keepdim=True)
+        x_se = self.fc(x_se)
+        return x * self.activate(x_se)
+
+
+class ESEAttn(nn.Module):
```
Comment (hhaAndroid) [inline mmyolo/models/layers/yolo_bricks.py:403]: ESEAttn 和 EffectiveSELayer 好奇怪呀？ 两者就差了一个 atten?

File: mmyolo/models/layers/yolo_bricks.py [hunk 43]
```diff
@@ -655,3 +761,255 @@ def forward(self, x) -> Tensor:
                 torch.cat([x1] + [m(x1) for m in self.poolings], 1))
         x2 = self.short_layers(x)
         return self.final_conv(torch.cat((x1, x2), dim=1))
+
+
+class SPP(nn.Module):
```
Comment (hhaAndroid) [inline mmyolo/models/layers/yolo_bricks.py:766]: PPOYLOESPPFBottleneck。 我们直接支持 fast 版本

Comment (hhaAndroid) [reply mmyolo/models/layers/yolo_bricks.py:766]: 这个可以直接复用我们已经实现的 SPPFBottleneck 吗

File: mmyolo/models/layers/yolo_bricks.py [hunk 44]
```diff
@@ -655,3 +761,255 @@ def forward(self, x) -> Tensor:
...
+                 in_channels: int,
+                 out_channels: int,
+                 num_block: int,
+                 block_cfg: ConfigType = dict(
+                     type='PPYOLOEBasicBlock', shortcut=True, use_alpha=True),
+                 stride: int = 1,
+                 norm_cfg: ConfigType = dict(
+                     type='BN', momentum=0.1, eps=1e-5),
+                 act_cfg: ConfigType = dict(type='SiLU', inplace=True),
+                 effective_se_cfg: OptMultiConfig = dict(
```
Comment (hhaAndroid) [inline mmyolo/models/layers/yolo_bricks.py:915]: attention_cfg

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**